### ISSUE TITLE
Python concludes instrument set has finished before device has set value

### ISSUE THREAD

**Issue by AdriaanRol** [microsoft/Qcodes#236]
A blocking get implemented by always calling get after set solves two problems 
1. Python concludes instrument set has finished before device has set value (on TCPIP, see below) 
2. Takes care of any rounding or converting an instrument does. 

I think this makes sense as a default behaviour for **any** set function. The parallel processing of the measurement loop should ensure the desired non-blocking behaviour for other cases and ensures proper waiting for completion before moving on to a critical step 
### Steps to reproduce
1. Call a set on an instrument controlled over TCPIP 
2. Get/measure some parameter on another device sufficiently fast 
3. In rare cases you will get the value before the first device has updated it's settings 

Note: this does not involve any parallel processing but it will also break loops using parallelism if they wait for a non-block condition. 
### Expected behaviour

The get-command should wait for the initial instrument to finish setting. 
### Actual behaviour

It waits for the command to finish (that is the command to be dumped in the TCPIP queue) but python concludes the command is completed before the action has been executed on the hardware. 
### Proposed solution

Add an argument to parameters (upon initialization) that is get_after_set. When this is True it will always call the get_command of the instrument after the set command. This will ensure the stored value in the log is correct (as sometimes instruments round values or do similar changes). 
I expect the wait for the response on the get will ensure the command is only considered complete after the device has updated it's setting (as it goes over the same communcations). 

This can be tested in a testsuite by having some virtual instrument in which there is an artificially long set-command and multiple communications to the virtual instrument. 
### System

**Winows 7**

**cf4d6f3**

**Comment by AdriaanRol** [comment]
@core Today I ran into this problem again. 
I have implemented a very simple snippet that makes a parameter always check execution. 

Could someone (most likely @alexcjohnson ) let me know what you think of this implementation? 

```python
class HandshakeParameter(StandardParameter):
    """
    If a string is specified as a set command it will append '*OPC?' and use
    instrument.ask instead of instrument.write
    """
    def _set_set(self, set_cmd, set_parser):
        exec_str = self._instrument.ask if self._instrument else None
        if isinstance(set_cmd, str):
            set_cmd += '\n*OPC?'
        self._set = Command(arg_count=1, cmd=set_cmd, exec_str=exec_str,
                            input_parser=set_parser, no_cmd_function=no_setter)

        self.has_set = set_cmd is not None
```

I don't like this implementation because it will **always** verify commands, leading to a potential slowdown. However it does solve the immediate problem in a "nice" way.

**Comment by giulioungaretti** [comment]
@AdriaanRol  @damazter  the current implementation is to use sleeps after every set (aka delay).

@AdriaanRol yes, I see your issue but how do you decide when or not to verify commands?

**Comment by giulioungaretti** [comment]
@AdriaanRol  does is depend  on the connectors I am not sure how it's different between TPC and GPIB  ?

**Comment by AdriaanRol** [comment]
@giulioungaretti . The downside of sleep is that it sets a fixed delay, there is no way to know if this delay is enough so you usually end up with a delay that is most of the time too long. This implementation asks the device if it has completed the command (as per standard SCPI standard). 

> @AdriaanRol yes, I see your issue but how do you decide when or not to verify commands?

You specifiy it when adding the parameter to an instrument. You either use StandardParameter (in which case you do not do this) or you use this subclass of the parameter.

**Comment by giulioungaretti** [comment]
@AdriaanRol so what you want is a blocking set

**Comment by AdriaanRol** [comment]
@giulioungaretti No. A blocking set is my immediate workaround for an issue I have now. 

What I want is an optionally blocking set, denoted by the use of keyword arguments in the set command as proposed by @damazter and me back in June.

**Comment by giulioungaretti** [comment]
@AdriaanRol the issue starts with this sentence:
"A blocking get implemented by always calling get after set solves two problems"
so I got confused by what is that is not right for your use. 

@AdriaanRol  and yes, I like that! We need that!
I would not want to ab-use kwargs for that but rather have a differently named function/method.
But  the question is if it has to be implemented with SCPI commands? or other ways

**Comment by AdriaanRol** [comment]
> @AdriaanRol the issue starts with this sentence:

The beauty of historical issues :) 
That was the original idea, discussion with @damazter led to the more general solution. 

Glad you like it :)

**Comment by giulioungaretti** [comment]
@AdriaanRol  got it!

**Comment by alexcjohnson** [comment]
@AdriaanRol Yes, I think this is a good solution. It only solves the first problem in the original statement, but I imagine the second one (reading back in any rounding or converting the instrument did) was a bit of a happy side-effect rather than the rationale.

Two possible concerns we may want to build into it:
- Will all instruments accept multiple commands at once like this? Probably, but this doesn't respect whatever terminator the instrument has other than `\n`. May want to instead make it two separate calls, one `_instrument.write` followed by an `_instrument.ask`, which you can either do here or maybe better by overriding `_validate_and_set` and `_validate_and_sweep` replacing `self._save_val` with this ask. If we did that, you'd also have the option to replace it with `self.get()` if that's what you really wanted.
- There will surely be instruments where you want to do this but they don't support `*OPC?` or it doesn't do what it's supposed to - there should be a way to specify an alternate handshake command.

**Comment by AdriaanRol** [comment]
@alexcjohnson I agree on both comments. I think allowing any function to be defined in the validate option is the way to go (possibly even specify both as a function and as a string). My example with `*OPC?`was just my initial solution. In any case OPC is supported by the SCPI standard so it should be fairly common. 

If you can implement a version that implements your suggestions (I'm quite wary of meddling with the instrument parameter), that would be great.

**Comment by giulioungaretti** [comment]
@AdriaanRol  let's wait for @alexcjohnson PR and documentation  before adding new APIs.

**Comment by giulioungaretti** [comment]
@AdriaanRol this an old one, but don't we think that instead of using the parameters //intrumetns to do this. It should be the loop responsibility to orchestrates the sequences of the actions ?
